project-docs/javascript-linter-spec.md

```diff
@@ -0,0 +1,163 @@
+- [javascript-constructor](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-constructor.yaml)
+- [javascript-method](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-method.yaml)
+- [javascript-property](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-property.yaml)
+- [javascript-error](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-error.yaml)
+- [javascript-language-feature](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-language-feature.yaml)
+
+## JavaScript recipe ingredients
+
+In total these recipes use the following ingredients:
+
+prose ingredients:
+
+- prose.short_description
+- prose.description?
+- prose.see_also
+- prose.syntax
+- prose.message
+- prose.error_type
+- prose.what_went_wrong
+- prose.\*
+
+data ingredients:
+
+- data.interactive_example?
+- data.class_constructor
+- data.constructor_properties?
+- data.static_properties?
+- data.static_methods?
+- data.instance_properties?
+- data.instance_methods?
+- data.examples
+- data.specifications
+- data.browser_compatibility
+
+All ingredients except two are demarcated in the page using an `<h2>` heading. The exceptions are `prose.short_description` and `data.interactive_example`.
+
+### Prose ingredients
+
+With three exceptions, all prose ingredients impose the same requirement on the page: it must have an `<h2>` section whose ID attribute maps to the named section. The mapping is simply that `prose.` is removed, and the first letter is capitalized:
+
+```
+prose.syntax      ->    Syntax
+prose.see_also    ->    See_also
+```
+
+The exceptions are `prose.short_description`, `prose.description?`, and `prose.*`.
+
+#### prose.short_description
+
+This ingredient isn't demarcated by an `<h2>` at all. To satisfy the ingredient, a page must:
+
+- contain a `<p>` element
+- that is not a note or a warning
+- before the interactive example or the first `<h2>`, whichever comes first
+
+#### prose.description?
+
+This is an optional prose ingredient. To satisfy the ingredient, if the page contains an `h2#Description`, then it must appear in the place specified by the recipe.
+
+#### prose.\*
+
+This ingredient enables writers to include extra sections, but ensuring that they always appear in the same place in the page. To satisfy this ingredient: if the page contains any `<h2>` headings that are not expected by the recipe, then they must appear at the point in the page that the `prose.*` ingredient is listed in the recipe.
+
+### Data ingredients
+
+Data ingredients typically impose more detailed requirements on pages than prose ingredients.
+
+#### data.interactive_example?
+
+This is an optional data ingredient. To satisfy the ingredient, if the page contains a call to the `{{EmbedInteractiveExample}}` macro, then the call must appear in the place specified by the recipe.
```

Comment:
This definition is short and extremely complicated. For most ingredients, I can take a slice of a page and tell you if it contains the ingredient (without a recipe, even). But I could never write a test for this ingredient alone.

I'm kind of uncomfortable with the idea that an ingredient is undefined until it's in a recipe. Do we lose something if we say that `data.interactive_example?` is an `{{EmbedInteractiveExample}}` macro call in a block by itself?

Reply:
The difficulty is that the ingredient is optional. So while "an `{{EmbedInteractiveExample}}` macro call in a block by itself" satisfies `data.interactive_example?`, so does "".

Sometimes (for example, for `data.static_methods`) we manage to distinguish three cases:

1. the page omits the ingredient (OK for optional ingredients)
2. the page tries to include the ingredient but specifies it wrong (not OK)
3. the page includes the ingredient and gets it right (OK for optional and mandatory ingredients)

Even there this is a hard line to maintain. If the page contains `H2#Static_Methods` is this case 1 or case 2? What about `H2#Static_functions`. And so on. Realistically, there's just going to be some stuff we get wrong.

But with `data.interactive_example?` it's hard to see any space to accommodate case 2.

> For most ingredients, I can take a slice of a page and tell you if it contains the ingredient (without a recipe, even). But I could never write a test for this ingredient alone.

This gets into the question above of whether the "?" is part of the ingredient or not. If it's not, then you can easily write a test for whether this section "contains the ingredient" - does it contain `{{EmbedInteractiveExample}}`? But when the ingredient is optional that's not a useful test. So it seems that the test for a recipe containing `data.interactive_example?` just collapses into the requirement for the ingredient (i.e. the macro call) to appear in the given place.

Reply:
Can we define the interactive example by its position in the page without reference to other ingredients? Since we only use interactive examples after short descriptions, we could bake that assumption into the ingredient. Something like this:

> To satisfy the ingredient, a page must have a a `<div>` containing an `{{EmbedInteractiveExample}}` macro call. The `<div>` must be immediately preceded by paragraph that is neither a note nor a warning. The `<div>` must precede any `<h2>`.

Reply:
This spec exists to serve us, so if that's useful I'm happy to add it. But although your suggestion does satisfy the ingredient, so does "the page does not contain a call to {{EmbedInteractiveExample}}". Would you be OK with:

> To satisfy the ingredient, either:
> * the page must not contain a `{{EmbedInteractiveExample}}` macro call, or
> * the page must have a `<div>` containing an `{{EmbedInteractiveExample}}` macro call. The `<div>` must be immediately preceded by paragraph that is neither a note nor a warning. The `<div>` must precede any `<h2>`.

Reply:
Yeah, this would work for me. Making, for example, an ingredient handler for this seems way easier than the one that has to know about the other ingredients to resolve. 👍

Reply:
OK -> https://github.com/mdn/stumptown-content/pull/391/commits/8222f7a2fa54b6eff281ad73ae2245fc048680e6